problems/en/04/README.md

```diff
@@ -0,0 +1,114 @@
+
+### Reading the logs 
+
+We can read our `feed` using `feed.createReadStream` and display that info on the console using `process.stdout`:
+
+```javascript
+feed.createReadStream().pipe(process.stdout)
+```
+
+As you can see, we are using the `pipe` method to _connect_ and define our data flow in our streams.
+
+```javascript
+// a --> b --> c
+// unix like: a | b | c
+a.pipe(b).pipe(c)
+```
+
+### Replication 
+
+Ok, let's suppose we have local feed and a remote public key. At some point we want to read data from this other feed, since we have it's PK, we can decrypt them.
+
+But before decrypt them, we need to fetch and merge them into our local feed. This process is called `replication`.
+
+![replicant scene from blade runner](https://media.giphy.com/media/xtpNfxNz7rTSo/giphy.gif)
+
+We want to _replicate_ remote feeds data into our local feed.
+
+In order to do this, we are going to use streams. Hypercore API has a `feed.replicate()` method which returns a _replication stream_ that reads the remote feed, sync it with the local feed and finally pass the result to the next stream. In other words, it behaves like a `DuplexStream`.
+
+![replicate](images/replicate.png)
+
+### Sync 
+
+With `replicate()` we can combine the _remote feed_ with our _local feed_ but we need to be aware of out of date data in our _remote feed_.
+
+!> Eventually, all the peers should have the same (up to date) data.
+
+If we see the conection between two peers as a bi-directional connection, we can do the following:
+```javascript
+//                (1)                      (2)
+remoteFeed.pipe(localFeed.replicate()).pipe(remoteFeed)
+```
+1. First, data is received from the remote feed and replicated into our local feed.
+2. Once our feed is updated, data is sent to the remote feed. This is made to ensure consistency.
+3. Finally, both feeds have the same data.
+
+## Exercise 
+
+We are going to simulate reading messages from another peer. To do that we need to:
+
+1. Sync local feed with remote one.
+2. Once sync is done, read data from our local feed and push each message into an array.
+3. When we finish reading our feed, we need to return the messages list.
+
+## Test
+
+```
+$ npm test ./04
+```
+
+## Tips
+
+### 1 - Pump
+
+By design, if we have streams connected by `.pipe` and one of them fails, the rest keep working.
+
+This can lead to multiple error conditions. We want to destroy all the streams if one fails. That's why we are going to use 
+[pump](/pump) instead of `.pipe`.
+
+!> As an extra feature, with `pump` we can pass a function to the last argument. This function will be executed when all the streams finish.
+
+```javascript
+a.pipe(b).pipe(c)
+
+// becomes
+
+pump(a, b, c, err => {
+  console.log('all streams have finished')
+})
+```
+
+### 2 - Reading/Writing data 
+
+A `WritableStream` can iterate through all the _chunks_ of data flowing in our streams and we can write them wherever we want, eg: filesystem, network, memory, etc. 
```

Comment:
I wouldn't add 'then' because it sounds like first the iteration happens, and then the writing happens, when in reality the writing is happening during the iteration.